drivers/SmartThings/matter-thermostat/src/init.lua

```diff
@@ -283,13 +474,153 @@ local function do_configure(driver, device)
   end
 end
 
-
 local function device_added(driver, device)
   device:send(clusters.Thermostat.attributes.ControlSequenceOfOperation:read(device))
   device:send(clusters.FanControl.attributes.FanModeSequence:read(device))
   device:send(clusters.FanControl.attributes.WindSupport:read(device))
 end
 
+local function store_unit_factory(capability_name)
+  return function(driver, device, ib, response)
+    device:set_field(capability_name.."_unit", ib.data.value, {persist = true})
+  end
+end
+
+local units = {
+  PPM = 0,
+  PPB = 1,
+  PPT = 2,
+  MGM3 = 3,
+  UGM3 = 4,
+  NGM3 = 5,
+  PM3 = 6,
+  BQM3 = 7,
+  PCIL = 0xFF -- not in matter spec
+}
+
+local unit_strings = {
+  [units.PPM] = "ppm",
+  [units.PPB] = "ppb",
+  [units.PPT] = "ppt",
+  [units.MGM3] = "mg/m^3",
+  [units.NGM3] = "ng/m^3",
+  [units.UGM3] = "μg/m^3",
+  [units.BQM3] = "Bq/m^3",
+  [units.PCIL] = "pCi/L"
+}
+
+local unit_default = {
+  [capabilities.carbonMonoxideMeasurement.NAME] = units.PPM,
+  [capabilities.carbonDioxideMeasurement.NAME] = units.PPM,
+  [capabilities.nitrogenDioxideMeasurement.NAME] = units.PPM,
+  [capabilities.ozoneMeasurement.NAME] = units.PPM,
+  [capabilities.formaldehydeMeasurement.NAME] = units.PPM,
+  [capabilities.veryFineDustSensor.NAME] = units.UGM3,
+  [capabilities.fineDustSensor.NAME] = units.UGM3,
+  [capabilities.dustSensor.NAME] = units.UGM3,
+  [capabilities.radonMeasurement.NAME] = units.BQM3,
+  [capabilities.tvocMeasurement.NAME] = units.PPM
+}
+
+-- All ConcentrationMesurement clusters inherit from the same base cluster definitions,
+-- so CarbonMonoxideConcentratinMeasurement is used below but the same enum types exist
+-- in all ConcentrationMeasurement clusters
+local level_strings = {
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.UNKNOWN] = "unknown",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.LOW] = "good",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.MEDIUM] = "moderate",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.HIGH] = "unhealthy",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.CRITICAL] = "hazardous",
+}
+
+local conversion_tables = {
+  [units.PPM] = {
+    [units.PPM] = function(value) return utils.round(value) end,
+    [units.UGM3] = function(value) return utils.round((value * 30.031 * 1000) / 24.45) end,
+  },
+  [units.PPB] = {
+    [units.PPM] = function(value) return utils.round(value/(10^3)) end
+  },
+  [units.PPT] = {
+    [units.PPM] = function(value) return utils.round(value/(10^6)) end
+  },
+  [units.MGM3] = {
+    [units.UGM3] = function(value) return utils.round(value * (10^3)) end,
+    [units.PPM] = function(value) return utils.round((value * 24.45) / (30.031)) end,
```

Comment:
What is the 24.45 and 30.031 value? These should be defined as constants somewhere

Also, I believe this is the molecular weight of formaldehyde, so these conversion would only work for formaldehyde measurements. We should find a way to generalize this so that this equation can be used for any concentration type. I would propose wrapping the conversion table in a function and then adding special handling for the cases where a molecular weight is required:


Here is some pseudo code:
```
local molecular_weights = {
  [formaldehyde] = 30.031
}

local function convert_unit(current_unit, target_unit, concentration_type)
  if current_unit or target_unit is one that needs to be converted (MGM3, UGM3, etc) then
    local molecular_weight = molecular_weights[concentration_type]
    local value = conversion_tables[current_unit][target_unit]
    value = value / or * molecular_weight depending on the target/current unit
  else
    return conversion_tables[current_unit][target_unit]
end
```

Reply:
I went a slightly different approach, but it should still work as intended. Let me know what you think.